### ISSUE TITLE
Clouddriver http_proxy

### ISSUE THREAD

**Issue by vijay-sapient** [GoogleCloudPlatform/spinnaker-for-gcp#148]
Hi Team, 

Is it possible to send http_proxy configurations for clouddriver. We are trying to connect to github for artefacts, and see that the requests are not going through proxy. Just to test we enabled NAT and its working with the NAT. In our environment we will not be using NAT, so would need a way to add proxy settings to clouddriver. We already tried this configuration, but not working https://www.spinnaker.io/setup/quickstart/faq/#i-want-to-run-a-spinnaker-service-clouddriver-echo-etc-behind-an-http-proxy-server

getting the below error, when i set JAVA_OPTS in clouddriver.yml

2019-11-11 00:22:56.303  WARN 1 --- [           main] c.n.s.c.a.gcs.GcsArtifactConfiguration   : Failure instantiating gcs artifact account GcsArtifactAccount(name=gcs-install-account, jsonPath=null): 

java.io.IOException: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
	at com.google.auth.oauth2.DefaultCredentialsProvider.getDefaultCredentials(DefaultCredentialsProvider.java:134) ~[google-auth-library-oauth2-http-0.18.0.jar:na]
	at com.google.auth.oauth2.GoogleCredentials.getApplicationDefault(GoogleCredentials.java:113) ~[google-auth-library-oauth2-http-0.18.0.jar:na]
	at com.google.auth.oauth2.GoogleCredentials.getApplicationDefault(GoogleCredentials.java:85) ~[google-auth-library-oauth2-http-0.18.0.jar:na]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactCredentials.<init>(GcsArtifactCredentials.java:61) ~[clouddriver-artifacts.jar:na]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactConfiguration.lambda$gcsArtifactCredentials$0(GcsArtifactConfiguration.java:47) ~[clouddriver-artifacts.jar:na]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193) ~[na:1.8.0_212]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382) ~[na:1.8.0_212]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[na:1.8.0_212]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[na:1.8.0_212]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_212]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_212]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_212]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactConfiguration.gcsArtifactCredentials(GcsArtifactConfiguration.java:54) ~[clouddriver-artifacts.jar:na]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactConfiguration$$EnhancerBySpringCGLIB$$f63aa27f.CGLIB$gcsArtifactCredentials$0(<generated>) ~[clouddriver-artifacts.jar:na]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactConfiguration$$EnhancerBySpringCGLIB$$f63aa27f$$FastClassBySpringCGLIB$$b08ac6c1.invoke(<generated>) ~[clouddriver-artifacts.jar:na]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at com.netflix.spinnaker.clouddriver.artifacts.gcs.GcsArtifactConfiguration$$EnhancerBySpringCGLIB$$f63aa27f.gcsArtifactCredentials(<generated>) ~[clouddriver-artifacts.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_212]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_212]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_212]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_212]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:607) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1321) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1160) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.addCandidateEntry(DefaultListableBeanFactory.java:1467) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1431) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveMultipleBeans(DefaultListableBeanFactory.java:1322) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1209) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1171) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:218) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1341) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1187) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:277) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1251) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1171) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:218) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1341) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1187) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141) ~[spring-boot-2.1.7.RELEASE.jar:2.1.7.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:743) ~[spring-boot-2.1.7.RELEASE.jar:2.1.7.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:390) ~[spring-boot-2.1.7.RELEASE.jar:2.1.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312) ~[spring-boot-2.1.7.RELEASE.jar:2.1.7.RELEASE]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:140) ~[spring-boot-2.1.7.RELEASE.jar:2.1.7.RELEASE]
	at org.springframework.boot.builder.SpringApplicationBuilder$run$0.call(Unknown Source) ~[na:na]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) ~[groovy-2.5.7.jar:2.5.7]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115) ~[groovy-2.5.7.jar:2.5.7]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:127) ~[groovy-2.5.7.jar:2.5.7]
	at com.netflix.spinnaker.clouddriver.Main.main(Main.groovy:78) ~[clouddriver-web.jar:na]

**Comment by demobox** [comment]
Thanks for raising this, Vijay! Do you by any chance know if this works for you when you're not using Spinnaker for GCP, e.g. if you have Spinnaker running in a GKE cluster installed in a different way (e.g. [using this codelab](https://codelabs.developers.google.com/codelabs/cloud-spinnaker-kubernetes-cd/#0)?

**Comment by vijay-sapient** [comment]
Hi Andrew, 

We were able to get the proxy working by adding below to ~/.hal/default/service-settings/clouddriver.yml

`env:
  JAVA_OPTS: "-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -XX:MaxRAMFraction=2 -Dhttp.proxyHost=PROXY_HOST -Dhttp.proxyPort=PROXY_PORT -Dhttps.proxyHost=PROXY_HOST -Dhttps.proxyPort=PROXY_PORT -Dhttp.nonProxyHosts='localhost|127.0.0.1|*.cluster.local|*.spinnaker|*.halyard'"
  HTTP_PROXY: "http://PROXY_HOST:PROXY_PORT"
  HTTPS_PROXY: "http://PROXY_HOST:PROXY_PORT"
  NO_PROXY: "localhost,127.0.0.1,*.cluster.local,*.spinnaker,*.halyard"
`
But the real issue was using this solution previously, i was able to add new GKE clusters within the same project using the kubeconfig file from my local machine and had no issues. I think some how, when clouddriver runs it rotates the keys.

But now when am trying to add GKE cluster which are from a different project. Clouddriver does not rotate the keys. So i followed https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/ and now creating token for authentication.

Not an expert, but i think there is a cron which runs to keep checking the kubernetes accounts and the time my token expires, clouddriver throws error saying it had problem to generating token, which glcoud generates with have service key admin access. This is works if the kubernetes account is within the same project, but not across project.
I would say its similar to workload identity, as we are only passing cluster info to hal

If am correct, it would be great if we could add functionality in clouddriver to rotate keys from different GCP projects.

To reproduce this, create a cluster on different project and add that cluster as a kubernetes account to cluster running spinnaker. once done generate the credentials and add it hal and deploy. you will see as soon as token expires clouddriver complains about problem in getting token.

**Comment by demobox** [comment]
> We were able to get the proxy working by adding below to ~/.hal/default/service-settings/clouddriver.yml

Ah, great, thanks for the update on that!

> I would say its similar to workload identity, as we are only passing cluster info to hal

Workload identity should be supported by Spinnaker for GCP now, if that would help in your situation - see #142 (and see [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#upgrade_the_management_console) for instructions on how to update Spinnaker for GCP's management console, if necessary).

As regards the GKE auth issue you're describing, how did you add the Spinnaker accounts to connect to the GKE clusters? Did you use the commands provided by the Spinnaker for GCP management console (described under "Add cloud provider accounts" [here](https://cloud.google.com/docs/ci-cd/spinnaker/spinnaker-for-gcp#manage_spinnaker)), or did you use a different approach?

**Comment by vijay-sapient** [comment]
Yes I ran ~/spinnaker-for-gcp/scripts/manage/add_gke_account.sh

to add account to Spinnaker

**Comment by demobox** [comment]
> Yes I ran ~/spinnaker-for-gcp/scripts/manage/add_gke_account.sh

OK, thanks for confirming!

**Comment by demobox** [comment]
Quick follow-up ping for @duftler here...

**Comment by duftler** [comment]
> you will see as soon as token expires clouddriver complains about problem in getting token.

@vijay-sapient How long after you first configure the new account do you see this expiration? Just trying to figure out what I would need to do to reproduce this.

**Comment by vijay-sapient** [comment]
exactly an hour, because gcloud get-credential creates the token which expires after an hour

**Comment by duftler** [comment]
Great. Will take following steps to try to reproduce:

1) Create a Spinnaker for GCP installation in a fresh project
2) Create a cluster in a different GCP project
2) Follow the instructions at https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/ to configure a k8s service account using a token, grant that k8s service account `cluster-admin` on the cluster in the other GCP project, and use `hal` to add it as a Spinnaker account
3) Push_and_apply the changes
4) Wait an hour and look for expiration messages

**Comment by duftler** [comment]
I followed the steps above, and was unable to reproduce. I can successfully deploy resources into the cluster in the other GCP project and I don't see any errors related to tokens expiring (waited more than 1 hour)/

Couple of things worth mentioning here:
- I created my kube config file (the one containing the credentials for accessing the cluster in the other GCP project) in `/home/duftler/.hal/default/credentials/my-kube-config`. Did this for the reasons described here: 
![image](https://user-images.githubusercontent.com/8437718/70657096-35cbb180-1c29-11ea-9a88-c90d7e613204.png)
- When following [these instructions](https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/), I changed the `hal config provider kubernetes account add my-test-account` command by dropping `--docker-registries my-gcr-account` and adding `--provider-version v2`. The command looked like this: `hal config provider kubernetes account add my-test-account --context $(kubectl config current-context) --provider-version v2 --kubeconfig-file ~/.hal/default/credentials/my-kube-config`
- If you want to be sure that your kube config file was specified and resolved correctly, wait until you've pushed and applied everything and then shell into your (newly-created) clouddriver pod and run `grep kubeconfigFile /opt/spinnaker/config/clouddriver.yml`. Then verify that the referenced file actually exists on the pod.
- If you are following the rbac-specific instructions you linked, the gcloud auth provider is not required. Once the `user.token` value is present in the kube config file for the service account user, you can remove the gcloud auth provider. If left configured, it tries to refresh the access token (i.e. `user.auth-provider.config.access-token`) once it has expired. If you have specified a `user.token`, you shouldn't need this other auth provider at all. I don't _think_ it will cause any problems if it is present, but `kubectl` may actually be trying all the auth schemes in parallel to see which one succeeds first (I don't know exactly how it works). If this is the case, maybe the gcloud auth provider is throwing an exception for you, but not causing other issues. The main thing to check is if you can actually deploy resources using that Spinnaker account.
- You can confirm that kubectl is truly making calls on clouddriver's behalf using the service account token by deleting the clusterrolebinding. You should (in less than a minute) start seeing error messages like this one from clouddriver (and the ui will no longer render resources deployed into that cluster):
```
2019-12-11 20:33:56.403  WARN 1 --- [cutionAction-20] c.n.s.c.cache.LoggingInstrumentation     : kubernetes:my-test-account/KubernetesCoreCachingAgent[1/1] completed with one or more failures

com.netflix.spinnaker.clouddriver.kubernetes.v2.op.job.KubectlJobExecutor$KubectlException: Failed to read [namespace] from null: Error from server (Forbidden): namespaces is forbidden: User "system:serviceaccount:default:spinnaker-service-account" cannot list resource "namespaces" in API group "" at the cluster scope

        at com.netflix.spinnaker.clouddriver.kubernetes.v2.op.job.KubectlJobExecutor.list(KubectlJobExecutor.java:451) ~[clouddriver-kubernetes-v2.jar:na]
        at com.netflix.spinnaker.clouddriver.kubernetes.v2.security.KubernetesV2Credentials.lambda$list$7(KubernetesV2Credentials.java:444) ~[clouddriver-kubernetes-v2.jar:na]
        at com.netflix.spinnaker.clouddriver.kubernetes.v2.security.KubernetesV2Credentials.runAndRecordMetrics(KubernetesV2Credentials.java:591) ~[clouddriver-kubernetes-v2.jar:na]
```
- Adding the clusterrolebinding back should cause those exceptions to no longer appear.

**Comment by vijay-sapient** [comment]
> * Follow the instructions at https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/ to configure a k8s service account using a token, grant that k8s service account `edit` on the cluster in the other GCP project, and use `hal` to add it as a Spinnaker account

I think, If you use these instructions, it's not workload identity as token are used and stored to authenticate. we should not be generating SERVICE_ACCOUNT_TOKEN.

**Comment by duftler** [comment]
Feels like multiple things are being conflated here. I was responding to the issue [here](https://github.com/GoogleCloudPlatform/spinnaker-for-gcp/issues/148#issuecomment-553171482) where tokens appeared to be expiring.

Workload identity is unrelated to this. Workload identity is used for consuming GCP services from GKE applications. We've been discussing and troubleshooting accessing GKE clusters.

If you want to use workload identity for its intended purpose, you can follow the links Andrew provided [here](https://github.com/GoogleCloudPlatform/spinnaker-for-gcp/issues/148#issuecomment-553185245).

**Comment by vijay-sapient** [comment]
@duftler from my previous comment 
"Yes I ran ~/spinnaker-for-gcp/scripts/manage/add_gke_account.sh

to add account to Spinnaker"

if we run ~/spinnaker-for-gcp/scripts/manage/add_gke_account.sh. The tokens will expire because add_gke_account.sh does not add tokens.

**Comment by duftler** [comment]
So the issue you are reporting is that accounts added via `add_gke_account.sh` stop working after 1 hour? And this happens when the cluster you are adding an account for is in another GCP project?

You mentioned following these instructions: https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/ so I did the same.

I can try to reproduce by using solely 'add_gke_acount.sh`, if that's what you are saying the current issue is.

**Comment by vijay-sapient** [comment]
yes thats correct. if you use add_gke_acount.sh it does not create the token which was generated using https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/.

**Comment by duftler** [comment]
They do 2 different things. It is either/or. That is, you should choose one approach or the other.

`add_gke_account.sh` works because it grants the GCP service account permissions in the managed project.

The instructions (https://www.spinnaker.io/setup/quickstart/halyard-gke-deploy-rbac/) that you've linked to again work because they result in permissions granted to a K8S service account, and that approach requires a particular type of token to authenticate as the k8s service account.

If either of these schemes works, the calls will succeed. These concepts are covered in detail in these public docs: https://cloud.google.com/kubernetes-engine/docs/how-to/role-based-access-control

I have now worked through each approach in detail, and am unable to reproduce the behavior you describe. I have left accounts configured via each scheme running for quite a few hours, and no exceptions are thrown.

Going to close this issue for now.

If you can clearly state exactly what you are trying to accomplish, I can try to provide guidance as to the best way to get there. If you want to do that, please do so in a new issue.

**Comment by duftler** [comment]
Have created an issue to improve the documentation for the `add_*_account` helper scripts in the console: https://github.com/GoogleCloudPlatform/spinnaker-for-gcp/issues/160
